voicevox_engine/guided/extractor.py

```diff
@@ -0,0 +1,98 @@
+import os
+from typing import List, Optional
+
+import numpy as np
+import pyworld as pw
+import snfa
+from scipy.signal import resample
+from snfa.viterbi import Segment
+
+from ..model import AudioQuery
+
+aligner: Optional[snfa.Aligner] = None
+
+
+def _lazy_init(model_path: str):
+    global aligner
+    if aligner is None:
+        if not os.path.exists(model_path):
+            raise Exception("No model exists, check the build pipeline")
+        aligner = snfa.Aligner(model_path)
+
+
+def _query2phoneme(query: AudioQuery) -> List[str]:
+    phoneme_list = []
+    for accent_phrase in query.accent_phrases:
+        for mora in accent_phrase.moras:
+            if mora.consonant is not None:
+                phoneme_list.append(mora.consonant)
+            phoneme_list.append(mora.vowel)
+        if accent_phrase.pause_mora:
+            phoneme_list.append("pau")
+    return ["pau"] + phoneme_list + ["pau"]
+
+
+def _align(wav: np.ndarray, src_sr: int, query: AudioQuery):
+    if len(wav.shape) == 2:
+        wav = np.sum(wav, axis=1) / 2
+    wav = resample(wav, aligner.sr * wav.shape[0] // src_sr)
+    ph = _query2phoneme(query)
+    segments, *_ = aligner(wav, ph, use_sec=False)
+
+    return segments
+
+
+def _norm_pitch(segments: List[Segment], pitch: np.ndarray):
+    dur_length = sum([seg.length for seg in segments])
+    scale_factor = pitch.shape[0] / dur_length
+    normed_pitch = []
+    for seg in segments:
+        normed_pitch.append(
+            np.average(
+                pitch[int(seg.start * scale_factor) : int(seg.end * scale_factor)]
+            )
```

Comment:
I got the impression that the pitch tends to be low at the end of sentences. I'm not entirely sure why this is happening, but it might be due to averaging in the unvoiced regions (where pitch is 0). How about calculating the average only over intervals where the pitch is greater than 0? Just make sure to account for cases where the pitch is 0 throughout.

Reply:
You're right, I modified the code to calculate only voiced phonemes on both estimated and predicted pitch, the result turns out to be pretty good:
[hiho_result.zip](https://github.com/VOICEVOX/voicevox_engine/files/12444979/hiho_result.zip)
I guess the rest work is test and ci workflow.